src/util/pg.py

```diff
@@ -196,27 +196,40 @@ def explode_query_range(cr, query, table, alias=None, bucket_size=10000, prefix=
 
     alias = alias or table
 
-    cr.execute("SELECT min(id), max(id) FROM {}".format(table))
-    min_id, max_id = cr.fetchone()
-    if min_id is None:
-        return []  # empty table
-
     if "{parallel_filter}" not in query:
         sep_kw = " AND " if re.search(r"\sWHERE\s", query, re.M | re.I) else " WHERE "
         query += sep_kw + "{parallel_filter}"
 
-    if ((max_id - min_id + 1) * 0.9) <= bucket_size:
-        # If there is less than `bucket_size` records (with a 10% tolerance), no need to explode the query.
-        # Force usage of `prefix` in the query to validate it correctness.
-        # If we don't the query may only be valid if there is no split. It avoid scripts to pass the CI but fail in production.
+    cr.execute(
+        format_query(
+            cr,
+            """
+            SELECT min(id), max(id)
+              FROM (SELECT id,
+                           (ROW_NUMBER() OVER (ORDER BY id) - 1) / %s
+                      FROM {table}
+                   ) t(id, g)
+             GROUP BY g
+             ORDER BY g
+            """,
+            table=table,
+        ),
+        [bucket_size],
+    )
+    if not cr.rowcount:
+        return []  # empty table
+
+    if cr.rowcount == 1:
+        # Only one bucket, no need to explode the query.
+        # Since the query may only be valid if there is no split, force usage of `prefix` in the query to validate its
+        # correctness and avoid scripts that pass the CI but fail in production.
```

Comment:
```suggestion
        # Since the query may only be valid if there is no split, force usage of `prefix` in the query to validate its
        # correctness and avoid scripts that pass the CI but fail in production.
```

I guess I'm lacking some context to understand this comment however I'm under the impression that the single-bucket case would be handled correctly without this if-branch (parallel_filter would be `BETWEEN <min(id)> AND <max(id)>` instead of `id IS NOT NULL`, I don't see a problem with that). Am I wrong?

Reply:
The idea is to ensure we validate the query always with the parallel filter. To avoid it failing later when we _actually_ try to use it.